internal/cluster/connection_manager.go

```diff
@@ -392,48 +390,42 @@ func (m *ConnectionManager) removeConnection(conn *Connection) {
 }
 
 func (m *ConnectionManager) tryConnectCluster(ctx context.Context) (pubcluster.Address, error) {
-	current := m.failoverService.Current()
-	m.logger.Trace(func() string {
-		return fmt.Sprintf("ConnectionManager: trying to connect to cluster: %s", current.ClusterName)
-	})
-	maxAttempts := math.MaxInt32
-	if m.failoverEnabled {
-		// when failover configs are provided, we want to switch between alternative clusters
-		maxAttempts = 1
-	}
-	// try the current cluster
-	addr, err := m.tryConnectCandidateCluster(ctx, current, maxAttempts)
-	if err == nil {
-		return addr, nil
-	}
-	// try all of the next alternative clusters
-	addr, ok := m.failoverService.TryNextCluster(func(next *CandidateCluster) (pubcluster.Address, bool) {
-		m.logger.Infof("trying to connect to next cluster: %s", next.ClusterName)
-		if addr, err := m.tryConnectCandidateCluster(ctx, next, maxAttempts); err == nil {
-			m.logger.Infof("successfully connected to cluster: %s", next.ClusterName)
-			return addr, true
+	tryCount := math.MaxInt64
+	if m.failoverConfig.Enabled {
+		tryCount = m.failoverConfig.TryCount
+	}
+	var nonRetryableErr cb.NonRetryableError
+	for i := 1; i <= tryCount; i++ {
+		next := m.failoverService.Next()
+		m.logger.Infof("trying to connect to cluster: %s", next.ClusterName)
+		addr, err := m.tryConnectCandidateCluster(ctx, next, next.ConnectionStrategy)
+		if err == nil {
+			m.logger.Infof("connected to cluster: %s", m.failoverService.Current().ClusterName)
+			return addr, nil
+		}
+		if nonRetryableErr.Is(err) || errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
+			break
 		}
-		return "", false
-	})
-	if ok {
-		return addr, nil
 	}
-	// notify if no successful cluster connection was established
 	return "", fmt.Errorf("cannot connect to any cluster: %w", hzerrors.ErrIllegalState)
 }
 
-func (m *ConnectionManager) tryConnectCandidateCluster(ctx context.Context, cluster *CandidateCluster, maxAttempts int) (pubcluster.Address, error) {
-	addr, err := m.cb.TryContext(ctx, func(ctx context.Context, attempt int) (interface{}, error) {
-		if attempt == maxAttempts {
-			m.logger.Debug(func() string {
-				return fmt.Sprintf("ConnectionManager: giving up connection attempts to cluster: %s", cluster.ClusterName)
-			})
-			return "", cb.WrapNonRetryableError(fmt.Errorf("giving up connection attempts to cluster: %w", hzerrors.ErrIllegalState))
-		}
+func (m *ConnectionManager) tryConnectCandidateCluster(ctx context.Context, cluster *CandidateCluster, cs *pubcluster.ConnectionStrategyConfig) (pubcluster.Address, error) {
+	tryCount := math.MaxInt32
+	if m.failoverConfig.Enabled {
+		tryCount = 1
```

Comment:
The old way was much more simpler to understand. 
I am against using CircuitBreaker for everything before we really decide to utilize its functionality. 

When I read this code it is hard to reason about. It feels like when failover config is not enabled we will try infinitely many but we are not. There is a second timeout parameter of CircuitBreaker which handles it.

Reply:
The circuit breaking behavior of CircuitBreakers in the code are not activated, so they just retry. Isn't it better to have a single way of having retry logic instead of repeating it throughout the code?

Reply:
Then lets make this part much more easier to read. 
How about if we have two kind of CircuitBreakers, one with the `tryCount` and other with `timeout` . 
And one can never set both of them. We don't necessarily create another class, I think we can easily have two flavors of constructors for CircuitBreaker. What do you think ?

Reply:
We come to the conclusion that we can not have two different constructors to achieve simplicity. 
Eventough it looks more complex to me comparing to what we have java , I will not insist on removing circuit breaker because of `a single way of having retry logic` argument.